### DISCUSSION TITLE
Modeling failed   Errors:     no valid days of data

### DISCUSSION THREAD

**Discussion by khris222** [USGS-R/streamMetabolizer#429] [Q&A] [answered]
Hi!
As a beginner of streammetabolizer, I am trying to calculate the metabolism of a stream using 24h continuous data. I read the quick start documentation and prepared my data based on the sampling data. 

Everything looks fine until the program runs the mm <- metab (). I always get this error:
Warning message:
In metab_fun(specs = specs, data = data, data_daily = data_daily,  :
  Modeling failed
  Errors:
    no valid days of data.

 I was confused since my data seems very similar to the sampling data after I convert solar.time into POSIXct according to the method proposed by Dr.Alison.(#172)

Here are the steps I've tried to do before, I would be very grateful if someone could answer my questions. And I have added my data in the attach files for viewing. Thanks a lot.
[jisuan.csv](https://github.com/DOI-USGS/streamMetabolizer/files/15137988/jisuan.csv)

dat<- read.csv('jisuan.csv')
dat$solar.time<- as.POSIXct(x=strptime(dat$solar.time,"%Y/%m/%d %H:%M"), tz="UTC")
library(unitted)
dat <-unitted(dat,c("","mgO2 L^-1","mgO2 L^-1","m", "degC", "umol m^-2 s^-1"))
library(streamMetabolizer)
library(dplyr)
bayes_name <- mm_name(type='bayes', pool_K600='none', err_obs_iid=TRUE, err_proc_iid=TRUE)
bayes_name
bayes_specs <- specs(bayes_name)
bayes_specs
mm <- metab(bayes_specs, data=dat)

By the way, the result of "bayes_specs" is as follows:
Model specifications:
  model_name               b_np_oipi_tr_plrckm.stan                                 
  engine                   stan                                                     
  split_dates              FALSE                                                    
  keep_mcmcs               TRUE                                                     
  keep_mcmc_data           TRUE                                                     
  day_start                4                                                        
  day_end                  28                                                       
  day_tests                full_day, even_timesteps, complete_data, pos_discharge...
  required_timestep        NA                                                       
  GPP_daily_mu             3.1                                                      
  GPP_daily_lower          -Inf                                                     
  GPP_daily_sigma          6                                                        
  ER_daily_mu              -7.1                                                     
  ER_daily_upper           Inf                                                      
  ER_daily_sigma           7.1                                                      
  K600_daily_meanlog       2.484906649788                                           
  K600_daily_sdlog         1                                                        
  err_obs_iid_sigma_scale  0.03                                                     
  err_proc_iid_sigma_scale 5                                                        
  params_in                GPP_daily_mu, GPP_daily_lower, GPP_daily_sigma, ER_dai...
  params_out               GPP, ER, DO_R2, GPP_daily, ER_daily, K600_daily, err_o...
  n_chains                 4                                                        
  n_cores                  4                                                        
  burnin_steps             500                                                      
  saved_steps              500                                                      
  thin_steps               1                                                        
  verbose                  FALSE    

This is my first time to post Q&A. Please forgive me if I disturb you. Thank you.

**Comment by robohall** [comment]
Check your time.  Problems often lie there. My sensors report in UTC (called Unix_Time) giving seconds since 1970.  These two lines of code will make the time needed for stream metabolizer.  Use package lubridate.

all_oxy$time<-as_datetime(all_oxy$Unix_Time)

all_oxy$solar.time<-convert_UTC_to_solartime(all_oxy$time, longitude= -113, time.type="mean solar”)

Solar time should be near you local time, but not exact unless your stream is in the longitudinal center of the time zone.

Looking at your data, I see highly variable O2 a good thing, and highly variable light, less good.  For the purposes of running sM I would use light generated by sM using only your locations and time.  Light your sensor sees is not necessarily light your stream sees and you will creat a lot of error as the stream GO chases this variability in light.

all_oxy$light<- calc_light(all_oxy$solar.time, latitude=48.5, longitude=-113, max.PAR =2326, attach.units = F)

Bob Hall

**Reply by khris222** [reply]
Hi Bob,

Thanks for your reply, very helpful information!

According to your suggestion, I recalculated solar.time and light. But the same mistake seems to have happened. I put the output data in the attach files.
[result.csv](https://github.com/DOI-USGS/streamMetabolizer/files/15140624/result.csv)

mm <- metab(bayes_specs, data=dat)
Warning message:
In metab_fun(specs = specs, data = data, data_daily = data_daily,  :
  Modeling failed
  Errors:
    no valid days of data

I have a guess whether this error is related to my failure to specify the desired resolution of the data in minutes (res) and day_start and day_end? Are these information necessary?

Also, is the max.PAR required to calculate light based on the maximum value of my PARlogger?

I really appreciate your reply, which is very helpful to me. I'm so sorry to disturb your weekend, have a nice day!

Niko

**Comment by robohall** [comment, marked as answer]
I see the problem.  You only sent one day of data.  A “day" in sM starts at 4 AM and goes to 03:59 the next day.  It ignore all data until it gets to 04:00, and then it needs a complete day after that .  You can override that in the specs file, but I would not start a sM model once the sun comes out.  More data.

Bob

**Reply by khris222** [reply]
Hi Bob,

Very helpful advice, problem solved! But I still have two more questions:

First, how to get the max.PAR used to calculate light. Is it the maximum value of the day in my PARlogger?

Second, I measured the metabolism of more than 50 streams, but for some reason a few streams only had O2 data from 7:00AM to 8:00AM(25 hours). According to your previous answer, are these streams unable to accurately estimate metabolism? Is it feasible to move the data after 7:00AM on the first day and before 4:00AM on the second day (7:00am-4:00am) to after 7:00AM on the second day in order to constitute a complete day?

Thank you so much for your answer, until yesterday I couldn't believe I was talking to Bob Hall. I am a graduate student and I am doing some research on stream metabolism. I have learned a lot of theoretical knowledge and got an idea from your manuscript before. I must commend you for your excellent work on stream metabolism. Your advice is very helpful to me. Thank you again Bob.

Niko

**Comment by robohall** [comment]
Niko,

You can use any number you want for Max PAR.  Because sM has a linear response to light it does not matter.  You could simply use 1.  Or 2000, which is typical noon PAR.  It only matters iif you use a non-linear light model in sM.


You might be fine with 7:00.  Give it a try, but update the specs file to start at the first time the sensor reads and then go 24 h.

Where are you going to school?

Bob

**Reply by khris222** [reply]
Hi Bob,

Thank you very much for your patient answer. Now my problem about sM has been solved.

I am a graduate student of the Institute of Hydrobiology, Chinese Academy of Sciences. I was very interested in stream metabolism, so I chose a related research topic. I often read your manuscripts to learn theoretical knowledge and the latest research progress.You really helped me a lot even though you may didn't know it. I have seen an interview about you and know that you like beer. Welcome to China to taste our beer and look forward to learning from you face to face.

cheers！

Niko
